### PR TITLE
Feature#481 resources needed tab

### LINKED ISSUE #481: WindowHutBuilder: material needed tab improvement
### Minecolonies version

Future version

### Expected behavior
Open the **resources needed** tab in the builder's hut gui:
- Have list of the needed resources with added informations: each item have it name display, the number of block available to the builder (i.e how many he have in his inventory + the chest's inventory) and the number needed. Example:
  - Coblestone 2/10 (color red)
  - Oak wood 0/10 (color red)
  - Oak wood plank 100/10 (color green)
- Click on a resource needed item to add it to the chest from the player inventory 
- Color coded item: 
  - green resource are already in the chest or builder inventory
  - red resource will need to be added to the chest if we want the build to be done 
- Order the list with the missing resources first to avoid constant scrolling
- Refresh the list every time we click on an item to reflect the changes

Difficulty: moderate 
Tag: enhancement

### Actual behaviour
- Go to the **resource needed** tab to have a look at what is missing
- switch to the chest to add the resources
- Close the gui
- Open the Builder gui again
- Navigate to the resource tab
- repeat and rinse until we have everything while memorizing what the builder have in his inventory and memorizing what the Builder chest have.

### PR DESCRIPTION
**Pull request by xavierh**
Closes #481

# Changes proposed in this pull request:
Open the resources needed tab in the builder's hut gui:

Have list of the needed resources with added informations: each item have it name display, the number of items missing for the builder (i.e how many are needed minus the one in his inventory - the chest's inventory), the number of items available and the number needed. Example:
- Cobblestone -2  8/10 (needed 10, have 8, need 2 more)
- Oak wood  -10 0/10 (needed 10 , have none, need 10 more)
- Oak wood plank +90  100/10 (need 10, have 100, have 90 to many )

Click on a resource needed item to add it to the chest from the player inventory
Additionally the resources are colour coded:
- Black: the builder have enough
- Red: the builder will need more of them
- Green: the builder will need more of them but the player have the correct amount in his inventory.

The list is ordered as follow:
- Green first;
- Red with button enabled (i.e. the player can give some of those items but not all of them)
- Red with button disabled (i.e. the player does not have any in his inventory)
- Black with button disabled (i.e. the builder have enough)

The list is refreshed when:
- The player is opening the gui.
- The player give some resources from the gui
- The chest or NPC inventory change
- The list is *not* refreshed when one of the additional chest have changed
-  Refresh the list every time we click on an item to reflect the changes

Review please

### CODE REVIEW

**Comment by Kostronor** [comment]
@xavierh please merge the branch back to your branch once :) I'll open a pull request for your convenience

File: src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java [hunk 1]
```diff
@@ -76,21 +135,57 @@ public int getElementCount()
...
             @Override
             public void updateElement(final int index, @NotNull final Pane rowPane)
             {
-                final Object obj = entries[index];
+                    final BuildingBuilder.BuildingBuilderResource resource = resources.get(index);
+                    Label resourceLabel = rowPane.findPaneOfTypeByID(RESOURCE_NAME, Label.class);
+                    Label neededLabel = rowPane.findPaneOfTypeByID(RESOURCE_AVAILABLE_NEEDED, Label.class);
+                    Button addButton = rowPane.findPaneOfTypeByID(RESOURCE_ADD, Button.class);
+
+                    switch (resource.getAvailabilityStatus())
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:146]: we can surely refactor this into it's own method

File: src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java [hunk 2]
```diff
@@ -76,21 +135,57 @@ public int getElementCount()
...
+                        case NOT_NEEDED:
+                        default:
+                            addButton.disable();
+                            resourceLabel.setColor(black,black);
+                            neededLabel.setColor(black,black);
+                            break;
+
+                    }
+
+                    //position the addRessource Button to the right
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:172]: this can also be its own method

**Comment by xavierh** [comment]
Not to sure what you mean. Do you mean that my last commit should be from
the merge?

File: src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java [hunk 3]
```diff
@@ -165,6 +176,95 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
...
+            HAVE_ENOUGHT
+        }
+
+        private String name;
+        private ItemStack itemStack;
+        private int amountAvailable;
+        private int amountNeeded;
+        private int amountPlayer;
+
+        BuildingBuilderResource(final String name, final ItemStack itemStack, final int amountAvailable, final int amountNeeded)
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:197]: modifier on this constructor and lots of docs for all public stuff missing

**Comment by marvin-bitterlich** [comment]
I overlooked the merge commit sorry

**Review by marvin-bitterlich — CHANGES_REQUESTED** [review]
This should definitely run through ci. @xavierh are you in @Minecolonies/developers ? if not please tell.
We change to target to an inhouse branch and repeat the pr to get ci feedback

**Comment by xavierh** [comment]
Actually I have done 2 commits since

**Comment by marvin-bitterlich** [comment]
closing to change target

**Pull request closed without merging**